cupy/sorting/search.py

```diff
@@ -181,7 +182,155 @@ def where(condition, x=None, y=None):
     return _where_ufunc(condition.astype('?'), x, y)
 
 
-# TODO(okuta): Implement searchsorted
+# This is to allow using the same kernels for all dtypes, ints & floats
+# as nan is a special case
+_preamble = '''
+template<typename T>
+__device__ bool _isnan(T val) {
+    return false;
+}
+template<>
+__device__ bool _isnan<float16>(float16 val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<float>(float val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<double>(double val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<const complex<double>&>(const complex<double>& val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<const complex<float>&>(const complex<float>& val) {
+    return isnan(val);
+}
+'''
+
+_searchsorted_kernel_left = core.ElementwiseKernel(
+    'S x, raw T bins, int32 n_bins',
+    'U y',
+    '''
+    y = 0;
+    if (_isnan<S>(x) || x > bins[n_bins-1]) {
+        y = n_bins;
+        return;
+    }
+    size_t l = 0;
+    size_t r = n_bins-1;
+    while (l<r) {
+        size_t m = l + (r - l) / 2;
+        if (bins[m] >= x) {
```

Comment:
By reversing this condition, you can easily see the only difference between the two kernels are operators `<=` and `<`.
How about writing that way, and possibly merging the implementations?

Reply:
I agree with this observation, however when adding `digitize` there is the need to check the direction of the array and the kernels have more differences.

I am fine with merging them now and re-split them later if we want to do a completely isolated PR.

Reply:
Thank you for explanation.
Let's leave the current implementation and review this point in another PR.